firedrake/adjoint_utils/blocks/assembly.py

```diff
@@ -10,8 +10,13 @@ class AssembleBlock(Block):
     def __init__(self, form, ad_block_tag=None):
         super(AssembleBlock, self).__init__(ad_block_tag=ad_block_tag)
         self.form = form
-        mesh = self.form.ufl_domain() if hasattr(self.form, 'ufl_domain') \
-            else None
+        if hasattr(self.form, 'ufl_domain'):
+            if isinstance(self.form, ufl.Interpolate):
+                mesh = ufl.domain.extract_unique_domain(self.form)
+            else:
+                mesh = self.form.ufl_domain()
```

Comment:
I think the correct thing should be something like `mesh = as_domain(form)`

Reply:
After we have the warning message
```
DeprecationWarning: Expr.ufl_domain() is deprecated, please use extract_unique_domain(expr) instead.
    warnings.warn("Expr.ufl_domain() is deprecated, please "
```
The function [ufl_domain](https://github.com/firedrakeproject/ufl/blob/master/ufl/core/expr.py#L260) returns `extract_unique_domain(self)`. So, I am using `mesh = ufl.domain.extract_unique_domain(self.form)` because that is used at [ufl_domain](https://github.com/firedrakeproject/ufl/blob/master/ufl/core/expr.py#L260).

Reply:
Do you think that is not the right way?